flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,211 @@
+
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.common.functions.RichMapPartitionFunction;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.ml.api.Estimator;
+import org.apache.flink.ml.common.datastream.DataStreamUtils;
+import org.apache.flink.ml.linalg.DenseMatrix;
+import org.apache.flink.ml.linalg.DenseVector;
+import org.apache.flink.ml.param.Param;
+import org.apache.flink.ml.util.ParamUtils;
+import org.apache.flink.ml.util.ReadWriteUtils;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An Estimator which implements the KNN algorithm.
+ *
+ * <p>See: https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm.
+ */
+public class Knn implements Estimator<Knn, KnnModel>, KnnParams<Knn> {
+
+    protected Map<Param<?>, Object> params = new HashMap<>();
+
+    public Knn() {
+        ParamUtils.initializeMapWithDefaultValues(params, this);
+    }
+
+    @Override
+    public KnnModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        /* Tuple2 : <sampleVector, label> */
+        DataStream<Tuple2<DenseVector, Double>> inputData =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                new MapFunction<Row, Tuple2<DenseVector, Double>>() {
+                                    @Override
+                                    public Tuple2<DenseVector, Double> map(Row value) {
+                                        Double label = (Double) value.getField(getLabelCol());
+                                        DenseVector feature =
+                                                (DenseVector) value.getField(getFeaturesCol());
+                                        return Tuple2.of(feature, label);
+                                    }
+                                });
+        DataStream<KnnModelData> distributedModelData = prepareModelData(inputData);
+        DataStream<KnnModelData> modelData = mergeModelData(distributedModelData);
+        KnnModel model = new KnnModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, getParamMap());
+        return model;
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return this.params;
```

Comment:
nits: Can we remove `this` here? 
In general we only use this.variable if the class member variable name collide with the function parameter name.

Also, could we rename `params` to `paramMap`? This could make the code style more consistent, but it is ok to me to leave as it is.

Reply:
OK